File: Algorithms/src/1339. Maximum Product of Splitted Binary Tree/Code/solution.js
var maxProduct = function (root) {
  const MOD = 1e9 + 7;
  let totalSum = 0;
  let maxProd = 0;

  function getTotalSum(node) {
    if (!node) return 0;
    return node.val + getTotalSum(node.left) + getTotalSum(node.right);
  }

  function dfs(node) {
    if (!node) return 0;

    let left = dfs(node.left);
    let right = dfs(node.right);

    let subtreeSum = node.val + left + right;
    maxProd = Math.max(maxProd, subtreeSum * (totalSum - subtreeSum));

    return subtreeSum;
  }

  totalSum = getTotalSum(root);
  dfs(root);

  return maxProd % MOD;
};

File: Algorithms/src/1339. Maximum Product of Splitted Binary Tree/Code/solution.py
class Solution:
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        MOD = 10**9 + 7
        self.max_prod = 0

        def get_total_sum(node):
            if not node:
                return 0
            return node.val + get_total_sum(node.left) + get_total_sum(node.right)

        def dfs(node):
            if not node:
                return 0

            left = dfs(node.left)
            right = dfs(node.right)

            subtree_sum = node.val + left + right
            self.max_prod = max(self.max_prod, subtree_sum * (total_sum - subtree_sum))

            return subtree_sum

        total_sum = get_total_sum(root)
        dfs(root)
        return self.max_prod % MOD
